File: new_practice/40.combination-sum-ii.py
#
# @lc app=leetcode id=40 lang=python3
#
# [40] Combination Sum II
#

# @lc code=start
class Solution:
    # def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        
    def combinationSum2(self, candidates, target): 
        res = []
        path = []
        candidates.sort()
        self.helper(candidates, target, path, res)

        return res




    def helper(self, candidates, target, path, res):
        # print(path)
        if target < 0: # wrong early return
            return

        if target == 0: # correct return
            if path not in res:
                res.append(path)
                return

        for idx in range(len(candidates)):
            ele = candidates[idx]
            # print(ele)
            self.helper(candidates[idx+1:], target-ele, path+[ele], res) 
            # candidates[idx+1:] 往後搜尋
            # path+[ele] 只有這次的ele, 改為list加入path








# @lc code=end

File: new_practice_cpp/40.combination-sum-ii.cpp
/*
 * @lc app=leetcode id=40 lang=cpp
 *
 * [40] Combination Sum II
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {

        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> path;

        dfs(candidates, target, ans, path, 0);

        return ans;        
    }

    void dfs(vector<int> &candidates, int target, vector<vector<int>> &ans,
             vector<int> &path, int begin){
        
        if(target == 0 && find(ans.begin(), ans.end(), path) == ans.end()){ // ans.end() means not found
            ans.push_back(path);
            return;
        }

        for(int i=begin; i<candidates.size() && target >= candidates[i] ;i++){
            path.push_back(candidates[i]);
            dfs(candidates, target-candidates[i], ans, path, i+1); // i+1 means next
            path.pop_back();
        }
    }
};
// @lc code=end

